objects/src/notes/note_execution_hint.rs

```diff
@@ -0,0 +1,189 @@
+                }
+                Ok(NoteExecutionHint::None)
+            },
+            1 => {
+                if payload != 0 {
+                    return Err(NoteError::InvalidNoteExecutionHintPayload(tag, payload));
+                }
+                Ok(NoteExecutionHint::Always)
+            },
+            2 => Ok(NoteExecutionHint::AfterBlock { block_num: payload }),
+            3 => {
+                let remainder = (payload >> 24 & 0xFF) as u8;
+                if remainder != 0 {
+                    return Err(NoteError::InvalidNoteExecutionHintPayload(tag, payload));
+                }
+
+                let epoch_len = ((payload >> 16) & 0xFF) as u8;
+                let slot_len = ((payload >> 8) & 0xFF) as u8;
+                let slot_offset = (payload & 0xFF) as u8;
+                let hint = NoteExecutionHint::OnBlockSlot { epoch_len, slot_len, slot_offset };
+
+                Ok(hint)
+            },
+            _ => Err(NoteError::InvalidNoteExecutionHintTag(tag)),
+        }
+    }
+
+    /// Returns whether the note execution conditions validate for the given `block_num`
+    ///
+    /// # Returns
+    /// - `None` if we don't know whether the note can be consumed.
+    /// - `Some(true)` if the note is consumable for the given `block_num`
+    /// - `Some(false)` if the note is not consumable for the given `block_num`
+    pub fn can_be_consumed(&self, block_num: u32) -> Option<bool> {
+        match self {
+            NoteExecutionHint::None => None,
+            NoteExecutionHint::Always => Some(true),
+            NoteExecutionHint::AfterBlock { block_num: hint_block_num } => {
+                Some(block_num >= *hint_block_num)
+            },
+            NoteExecutionHint::OnBlockSlot { epoch_len, slot_len, slot_offset } => {
+                let epoch_len_blocks: u32 = 1 << epoch_len;
+                let slot_len_blocks: u32 = 1 << slot_len;
+
+                let block_epoch_index = block_num / epoch_len_blocks;
+
+                let slot_start_block =
+                    block_epoch_index * epoch_len_blocks + (*slot_offset as u32) * slot_len_blocks;
+                let slot_end_block = slot_start_block + slot_len_blocks;
+
+                let can_be_consumed = block_num >= slot_start_block && block_num < slot_end_block;
+                Some(can_be_consumed)
+            },
+        }
+    }
+
+    pub fn into_parts(&self) -> (u8, u32) {
+        match self {
+            NoteExecutionHint::None => (0, 0),
+            NoteExecutionHint::Always => (1, 0),
+            NoteExecutionHint::AfterBlock { block_num } => (2, *block_num),
+            NoteExecutionHint::OnBlockSlot { epoch_len, slot_len, slot_offset } => {
+                let payload: u32 =
+                    ((*epoch_len as u32) << 16) | ((*slot_len as u32) << 8) | (*slot_offset as u32);
+                (3, payload)
+            },
+        }
+    }
+}
+
+#[cfg(test)]
```

Comment:
nit: I would add a section separator above this line.

Reply:
Done